File: Polynomial Regression.py
# Polynomial Regression.

# Import libraries.
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Import dataset.
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Split dataset into training set and test set.
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)"""

""" Feature scaling.
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)"""

# Fit linear regression model to dataset.
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Fit polynomial regression model to dataset.
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

# Visualizing linear regression results.
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Truth or Bluff (Linear Regression Results)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Visualizing polynomial regression results.
    # X_grid = np.arange(min(X), max(X), 0.1)
    # X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg_2.predict(poly_reg.fit_transform(X)), color = 'blue')
    # plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Regression Results)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Predict new result with linear regression.
lin_reg.predict(6.5)

# Predict new result with polynomial regression.
lin_reg_2.predict(poly_reg.fit_transform(6.5))

File: Polynomial Regression.r
# Polynomial Regression.
# Import dataset.
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting dataset into training set and test set.
# install.packages('catools')
# library(catools)
# set.seed(123)
# split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling.
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fit linear regression to dataset.
lin_reg = lm(formula = Salary ~ ., data = dataset)

# Fit polynomial regression to dataset.
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ ., data = dataset)

# Visualize linear regression results.
install.packages('ggplot2')
library(ggplot2)
ggplot() + geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = 'red') +
geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = )), colour = 'blue') +
ggtitle('Truth or Bluff (Linear Regression)') +
xlab('Level') +
ylab('Salary')

# Visualize polynomial regression results.
install.packages('ggplot2')
library(ggplot2)
ggplot() + geom_point(aes(x = dataset$Level, y = dataset$Salary), colour = 'red') +
geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = )), colour = 'blue') +
ggtitle('Truth or Bluff (Polynomial Regression)') +
xlab('Level') +
ylab('Salary')

# Predict new result with linear regression.
y_pred = predict(lin_reg, data.frame(Level = 6.5))

# Predict new result with polynomial regression.
y_pred = predict(poly_reg, data.frame(Level = 6.5, Level2 = 6.5^2, Level3 = 6.5^3, Level4 = 6.5^4))
